themes/qgrid/source/js/menu.js

```diff
@@ -15,26 +18,229 @@ function toggleVisibility(e) {
 	}
 }
 
-function filter(e) {
+function onSearchChange(e) {
 	e.stopImmediatePropagation();
 	const input = e.target || e.srcElement;
-	const filter = input.value.toLowerCase();
+	const search = input.value.toLowerCase();
+	setSearch(search);
+}
+
+function searchOnLoad() {
+	const active = document.querySelector('.active-topic');
+	const search = getSearch();
+
+	if (active && search) {
+		document.getElementById('search').value = search;
+		currentPageSearch(search);
+		menuItemsSearch(search);
+	}
+}
+
+function searchOnHashChange() {
+	const search = getSearch();
+	document.getElementById('search').value = search;
+
+	if (!search) {
+		document.getElementById('search').focus();
+	}
+	currentPageSearch(search);
+	menuItemsSearch(search);
+}
+
+function currentPageSearch(search) {
+	let scrolled = false;
+
+	document.querySelectorAll('.page-wrapper *').forEach(node => {
+		if (SEARCHEABLE_ELEMENTS.includes(node.nodeName)) {
+			const a = node.querySelectorAll('a');
+			const pageItem = (a[a.length - 1] && a[a.length - 1].textContent) ? a[a.length - 1] : node;
+			const { textContent } = pageItem;
+			const searchWords = splitToWords(search);
+			const firstMatch = (testSearch(textContent, search)) ? searchWords[indexOfSearch(textContent, searchWords)] : '';
+			const searchValue = (textContent.toLowerCase().includes(search)) ? search : firstMatch;
+
+			if (highlightText(pageItem, searchValue) && !scrolled) {
+				scrolled = true;
+				pageItem.scrollIntoView();
+			}
+		}
+	});
+}
+
+function menuItemsSearch(search) {
 	const nav = document.getElementById('nav');
-	const li = nav.getElementsByTagName('li');
-
-	for (let i = 0; i < li.length; i++) {
-		const a = li[i].getElementsByTagName('a')[0];
-		if (a) {
-			if (a.text.toLowerCase().indexOf(filter) > -1) {
-				li[i].style.display = '';
-			} else {
-				li[i].style.display = 'none';
+	const titles = nav.querySelectorAll('.title');
+
+	for (let title of titles) {
+		const menuItem = title.parentElement;
+		for (let tag of getTags(menuItem)) {
+			updateMenuLinks(title, tag, search);
+
+			if (testSearch(tag, search)) {
+				appendTagText(menuItem, tag, search);
+				const menuTag = menuItem.querySelector('.tag');
+				const { textContent } = menuTag;
+				const searchWords = splitToWords(search);
+				const firstMatch = searchWords[indexOfSearch(textContent, searchWords)];
+				const searchValue = (textContent.toLowerCase().includes(search)) ? search : firstMatch;
+				highlightText(menuTag, searchValue);
+				break;
 			}
+			removeTagText(menuItem, search);
 		}
+		highlightText(title, search);
 	}
 }
 
+function testSearch(tag, search) {
+	const searchWords = splitToWords(search);
```

Comment:
no need to make so explicit in internal variables if name is not intersecting with other local names
words will be enough here